File: app/routes/return_order_notice.py
from flask import Blueprint, request, render_template, jsonify, current_app
from config import config  # 导入配置实例
from app.utils.rabbitmq import push_message  # 复用RabbitMQ推送工具
import logging
import json
from datetime import datetime

logger = logging.getLogger(__name__)

# ==================== 蓝图定义 ====================
return_order_notice_bp = Blueprint('return_order_notice', __name__, url_prefix='/return_order_notice')


# ==================== 路由函数 ====================
# 通知单入库页面（GET请求）
@return_order_notice_bp.route('/')
def index():
    """通知单入库页面"""
    return render_template('return_order_notice.html', preset=config.RETURN_ORDER_NOTICE_PRESET)


# 通知单入库接口（POST请求）
@return_order_notice_bp.route('/submit', methods=['POST'])
def submit():
    """通知单入库接口"""
    try:
        # 1. 获取用户输入（基础字段）
        return_order_code = request.form.get('returnOrderCode')
        close_status = request.form.get('CloseStatus')
        warehouse_code = request.form.get('warehouseCode')
        detail_count = int(request.form.get('detail_count', 1))

        # 2. 验证必填字段
        required_fields = {
            'returnOrderCode': return_order_code,
            'warehouseCode': warehouse_code
        }
        
        missing_fields = [field for field, value in required_fields.items() if not value]
        if missing_fields:
            return jsonify({
                'status': 'error',
                'message': f'缺少必填字段: {missing_fields}'
            })
        
        # 处理CloseStatus字段，如果为空则赋值为空字符串
        close_status = close_status if close_status else ''

        # 3. 获取商品明细
        order_lines = []
        for i in range(1, detail_count + 1):
            item_code = request.form.get(f'itemCode{i}')
            actual_qty = request.form.get(f'actualQty{i}')
            
            # 验证明细必填字段
            if not item_code or not actual_qty:
                return jsonify({
                    'status': 'error',
                    'message': f'第{i}行明细缺少必填字段: 商品编码或数量'
                })
            
            order_lines.append({
                'itemCode': item_code,
                'actualQty': actual_qty,
                'inventoryType': 'ZP',
                'orderLineNo': str(i),
                'ownerCode': 'XIER'
            })

        # 4. 合并参数
        current_time = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        
        final_params = {
            'type': 2,
            'returnOrderCode': return_order_code,
            'callbackResponse': {
                'apiMethodName': 'returnorder.confirm',
                'orderLines': order_lines,
                'extendProps': {
                    'CloseStatus': close_status,
                    'ApiSource': 'FLUXWMS'
                },
                'responseClass': 'com.qimen.api.response.ReturnorderConfirmResponse',
                'returnOrder': {
                    'orderConfirmTime': current_time,
                    'orderType': 'THRK',
                    'outBizCode': '',
                    'ownerCode': 'XIER',
                    'remark': '',
                    'returnOrderCode': return_order_code,
                    'warehouseCode': warehouse_code
                },
                'version': '2.0'
            }
        }

        # 5. 推送消息到RabbitMQ
        queue_name = current_app.config.get('RETURN_ORDER_NOTICE_QUEUE', 'sale_return_plan_add_back_b2c')
        push_message(queue_name, final_params)

        return jsonify({
            'status': 'success',
            'message': '通知单入库成功',
            'data': final_params
        })
    except Exception as e:
        return jsonify({
            'status': 'error',
            'message': f'处理失败: {str(e)}'
        })

File: app/static/js/return_order_notice.js
/*
 * 通知单入库页面JavaScript
 * 实现页面加载时自动预览JSON和实时更新功能
 */

class ReturnOrderNoticeManager {
    constructor() {
        this.form = document.getElementById('notice-form');
        this.detailCountInput = document.getElementById('detail-count');
        this.detailFieldsContainer = document.getElementById('detail-fields');
        this.detailCountBadge = document.getElementById('detail-count-badge');
        this.jsonPreview = document.getElementById('json-preview');
        this.previewBtn = document.getElementById('preview-btn');
        this.resetBtn = document.getElementById('reset-btn');
        this.copyJsonBtn = document.getElementById('copy-json');
        this.expandJsonBtn = document.getElementById('expand-json');

        this.successModal = new bootstrap.Modal(document.getElementById('successModal'));
        this.errorModal = new bootstrap.Modal(document.getElementById('errorModal'));
        this.errorMessage = document.getElementById('error-message');

        this.preset = {
            returnOrderCode: 'RN202410010001',
            warehouseCode: 'WH001',
            CloseStatus: '',
            itemCode: 'ITEMXXX',
            actualQty: '10'
        };

        this.init();
    }

    init() {
        this.bindEvents();
        this.generateDetailFields();
        this.updateJsonPreview();
    }

    bindEvents() {
        // 明细数量变化
        this.detailCountInput.addEventListener('input', () => {
            const count = parseInt(this.detailCountInput.value, 10);
            if (count >= 1 && count <= 10) {
                this.generateDetailFields();
                this.updateJsonPreview();
            } else if (count < 1) {
                this.detailCountInput.value = 1;
                this.generateDetailFields();
                this.updateJsonPreview();
            } else if (count > 10) {
                this.detailCountInput.value = 10;
                this.generateDetailFields();
                this.updateJsonPreview();
            }
        });

        // 预览按钮
        this.previewBtn.addEventListener('click', () => {
            this.updateJsonPreview();
            this.showToast('数据预览已更新', 'info');
        });

        // 重置按钮
        this.resetBtn.addEventListener('click', () => {
            this.resetForm();
        });

        // 复制JSON
        this.copyJsonBtn.addEventListener('click', () => {
            this.copyJson();
        });

        // 展开JSON
        this.expandJsonBtn.addEventListener('click', () => {
            this.toggleJsonExpansion();
        });

        // 实时更新JSON预览
        this.form.addEventListener('input', () => {
            this.debounce(() => this.updateJsonPreview(), 500);
        });

        // 提交表单
        this.form.addEventListener('submit', (e) => this.handleSubmit(e));
    }

    generateDetailFields() {
        const count = parseInt(this.detailCountInput.value, 10) || 1;
        const fieldsConfig = JSON.parse(this.detailCountInput.getAttribute('data-fields') || '[]');

        this.detailFieldsContainer.innerHTML = '';
        this.detailCountBadge.textContent = `${count} 项`;

        for (let i = 1; i <= count; i++) {
            const detailRow = document.createElement('div');
            detailRow.className = 'row g-3 mb-3 pb-3 border-bottom';
            detailRow.innerHTML = `<div class="col-md-12"><h6 class="text-muted mb-2">明细 ${i}</h6></div>`;

            fieldsConfig.forEach(field => {
                const fieldCol = document.createElement('div');
                fieldCol.className = 'col-md-6';

                let inputHtml = `<input type="${field.type || 'text'}" class="form-control" `;
                if (field.min) inputHtml += `min="${field.min}" `;
                inputHtml += `id="${field.idPrefix}${i}" name="${field.namePrefix}${i}" placeholder="${field.placeholder}" ${field.required ? 'required' : ''}>`;

                fieldCol.innerHTML = `
                    <label for="${field.idPrefix}${i}" class="form-label">
                        ${field.label} ${field.required ? '<span class="text-danger">*</span>' : ''}
                    </label>
                    <div class="input-group">
                        <span class="input-group-text">
                            <i class="${field.icon}"></i>
                        </span>
                        ${inputHtml}
                    </div>
                `;

                detailRow.appendChild(fieldCol);
            });

            this.detailFieldsContainer.appendChild(detailRow);
        }
    }

    updateJsonPreview() {
        // 收集表单数据
        const formData = new FormData(this.form);
        const formValues = Object.fromEntries(formData.entries());
        const detailCount = parseInt(formValues.detail_count || 1, 10);

        // 使用预制参数填充缺失字段
        const returnOrderCode = formValues.returnOrderCode || this.preset.returnOrderCode;
        const warehouseCode = formValues.warehouseCode || this.preset.warehouseCode;
        const closeStatus = formValues.CloseStatus || this.preset.CloseStatus;

        // 收集明细数据，使用预制参数填充缺失值
        const orderLines = [];
        for (let i = 1; i <= detailCount; i++) {
            const itemCode = formValues[`itemCode${i}`] || `${this.preset.itemCode}${i}`;
            const actualQty = formValues[`actualQty${i}`] || this.preset.actualQty;

            orderLines.push({
                itemCode: itemCode,
                actualQty: actualQty,
                inventoryType: 'ZP',
                orderLineNo: i.toString(),
                ownerCode: 'XIER'
            });
        }

        // 获取当前时间
        const now = new Date();
        const year = now.getFullYear();
        const month = String(now.getMonth() + 1).padStart(2, '0');
        const day = String(now.getDate()).padStart(2, '0');
        const hours = String(now.getHours()).padStart(2, '0');
        const minutes = String(now.getMinutes()).padStart(2, '0');
        const seconds = String(now.getSeconds()).padStart(2, '0');
        const currentTime = `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;

        // 构建最终JSON
        const finalJson = {
            type: 2,
            returnOrderCode: returnOrderCode,
            callbackResponse: {
                apiMethodName: 'returnorder.confirm',
                orderLines: orderLines,
                extendProps: {
                    CloseStatus: closeStatus,
                    ApiSource: 'FLUXWMS'
                },
                responseClass: 'com.qimen.api.response.ReturnorderConfirmResponse',
                returnOrder: {
                    orderConfirmTime: currentTime,
                    orderType: 'THRK',
                    outBizCode: '',
                    ownerCode: 'XIER',
                    remark: '',
                    returnOrderCode: returnOrderCode,
                    warehouseCode: warehouseCode
                },
                version: '2.0'
            }
        };

        this.jsonPreview.textContent = JSON.stringify(finalJson, null, 2);
    }

    handleSubmit(e) {
        e.preventDefault();

        // 使用Fetch API提交表单
        fetch('/return_order_notice/submit', {
            method: 'POST',
            body: new FormData(this.form)
        })
        .then(response => response.json())
        .then(data => {
            if (data.status === 'success') {
                // 更新JSON预览
                this.updateJsonPreview();
                this.successModal.show();
            } else {
                this.errorMessage.textContent = data.message;
                this.errorModal.show();
            }
        })
        .catch(error => {
            this.errorMessage.textContent = `提交失败: ${error.message}`;
            this.errorModal.show();
        });
    }

    copyJson() {
        try {
            const textArea = document.createElement('textarea');
            textArea.value = this.jsonPreview.textContent;
            document.body.appendChild(textArea);
            textArea.select();
            document.execCommand('copy');
            document.body.removeChild(textArea);
            this.showToast('JSON已复制到剪贴板', 'success');
        } catch (error) {
            console.error('复制失败:', error);
            this.showToast('复制失败，请手动复制', 'error');
        }
    }

    toggleJsonExpansion() {
        const currentText = this.jsonPreview.textContent;
        try {
            const jsonObj = JSON.parse(currentText);
            if (currentText.includes('\n')) {
                // 折叠
                this.jsonPreview.textContent = JSON.stringify(jsonObj);
                this.expandJsonBtn.innerHTML = '<i class="fas fa-expand me-1"></i>展开';
            } else {
                // 展开
                this.jsonPreview.textContent = JSON.stringify(jsonObj, null, 2);
                this.expandJsonBtn.innerHTML = '<i class="fas fa-compress me-1"></i>折叠';
            }
        } catch (e) {
            // 不是有效的JSON，不做处理
        }
    }

    resetForm() {
        this.form.reset();
        this.detailCountInput.value = 1;
        this.generateDetailFields();
        this.updateJsonPreview();
        this.showToast('表单已重置', 'info');
    }

    debounce(func, wait) {
        clearTimeout(this.debounceTimer);
        this.debounceTimer = setTimeout(func, wait);
    }

    showToast(message, type = 'info') {
        // 创建toast元素
        const toastContainer = document.createElement('div');
        toastContainer.className = `toast fade show position-fixed top-20 end-0 p-3 bg-${type} text-white z-50`;
        toastContainer.setAttribute('role', 'alert');
        toastContainer.setAttribute('aria-live', 'assertive');
        toastContainer.setAttribute('aria-atomic', 'true');
        toastContainer.style.minWidth = '250px';

        toastContainer.innerHTML = `
            <div class="toast-header bg-${type} text-white border-0">
                <strong class="me-auto">提示</strong>
                <button type="button" class="btn-close btn-close-white" data-bs-dismiss="toast" aria-label="Close"></button>
            </div>
            <div class="toast-body">
                ${message}
            </div>
        `;

        document.body.appendChild(toastContainer);

        // 自动关闭
        setTimeout(() => {
            toastContainer.classList.remove('show');
            toastContainer.classList.add('hide');
            setTimeout(() => {
                document.body.removeChild(toastContainer);
            }, 500);
        }, 3000);
    }
}

// 确保Bootstrap已加载
if (typeof bootstrap === 'undefined') {
    console.error('Bootstrap is not loaded!');
}